http/README.md

```diff
@@ -0,0 +1,113 @@
+in the future (with a different `www-authenticate` value) or be added to the
+application protocols themselves.
+
+### Authentication flow
+
+1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `www-authenticate: libp2p-noise` (The server MAY also include `libp2p-token` as an authentication scheme).
+2. The client sets the `Authorization` [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) to `libp2p-noise <multibase-encoded-noise-protobuf>` . This initiates the `IX` or `NX` handshake.
+    1. The protobuf is multibase encoded, but clients MUST only use encodings that are HTTP header safe (refer to to the [token68 definition](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.2)). To set the minimum bar for interoperability, clients and servers MUST support base32 encoding (”b” in the multibase table).
+    2. When the server receives this request and `IX` was used, it can authenticate the client.
+3. The server responds with `Authentication-Info` field set to `libp2p-noise <multibase-encoding-noise-protobuf-response>`.
+    1. The server MUST include the SNI used for the connection in the [Noise extensions](https://github.com/libp2p/specs/blob/master/noise/README.md#noise-extensions).
+    2. The server MAY include a token that the client can use to avoid doing another Noise handshake in the future. The client would use this token by setting the `Authorization` header to `libp2p-token <token>`.
+    3. When the client receives this response, it can authenticate the server’s peer ID.
+4. The client verifies the SNI in the Noise extension matches the one used to initiate the connection. The client MUST close the connection if they differ.
+    1. The client SHOULD remember this connection is authenticated.
+    2. The client SHOULD use the `libp2p-token` if provided for future authorized requests.
+
+This costs one round trip, but can piggy back on an appropriate request.
+
+### Authentication Endpoint
+
+Because the client needs to make a request to authenticate the server, and the client may not want to make the real request before authenticating the server, the server MAY provide an authentication endpoint. This authentication endpoint is like any other application protocol, and it shows up in `.well-known/libp2p`, but it only does the authentication flow. It doesn’t send any other data besides what is defined in the above Authentication flow. The protocol id for the authentication endpoint is `/http-noise-auth/1.0.0`.
+
+## Using HTTP semantics over stream transports
+
+Application protocols using HTTP semantics can run over any libp2p stream transport. Clients open a new stream using `/http/1.1` as the protocol identifer. Clients encode their HTTP request as an HTTP/1.1 message and send it over the stream. Clients parse the response as an HTTP/1.1 message and then close the stream.
```

Comment:
This impose strong restriction on the implementations, this makes it way harder to use multiple HTTP clients in one single app.
To speak about what I know about in golang there are two main http implementations used, [`net/http`](https://pkg.go.dev/net/http) and [`fasthttp`](https://pkg.go.dev/github.com/valyala/fasthttp).
Some frameworks like [Fiber](https://gofiber.io/) are only compatible with `fasthttp`.

After talking with @MarcoPolo it seems that the expected minimum API for implementations would be along theses lines:
```go
func (*Host) SetHTTPHandler(protocol.ID, func(remote peer.ID, res ResponseWriter,  req *Request))
```
For equivalence in other languages, that means that libp2p would handle and or delegate deserializing the HTTP requests into a an http request and response writer objects before executing a callback into the consumer's code. This would lock ourself in the `net/http` ecosystem.
Even tho this API is convenient and I do want it in `go-libp2p` I think this spec should target implementation APIs at even lower level where `io.Reader` and `io.Writer` (or equivalent in other programing languages) are used instead.

> Note: Go is a language with an extremely strong std influence, the fact that multiple http libraries are exists and are used in production in Go lead me to question what would happen in a language like Rust or C++ which do not even http libraries in their stds.

Using a single stream makes that way harder to do because you don't know where the request should be routed without parsing it first.
The most reasonable way to solve this for io based routers with a single stream is to write a minimum HTTP parser which parse the path, find out where the request should be routed and then forward the request to the io based consumer with the already parsed buffer and the remaining stream.

However this does not solve a case where I have different applications supporting different versions.
Let's consider the case where I use `h2c` for `/ipfs-gateway-api` (the header compression helps reducing the request size for such a header chatty API or something idk this is an example).
And I also serve `/my-super-application` by a C application which is implemented by using pipes and fork+execve into some handler program, again this is just an example where I would need to use to pass the underying streams.
I can't say to other peers that I'll accept `h2c` but only for `/ipfs-gateway-api`, I have two solutions:
- If I settle for the lowest common denominator. This is bad for my `/ipfs-gateway-api` requests which now run over `http/1.1` and don't get header compression.
- If I advertise `http/h2c` too, what happen if I receive an `h2c` request for `/my-super-application`, I now need to transcode it to `http/1.1` adding lots of complexity.

> Note: I know `h2c` has an upgrade mechanism from `http/1.1` but it does not solve the problem for requests transitioning in the middle of a stream that is being reused (afaik you can't downgrade from `h2c` to `http/1.1` based on the paths you red if you don't like it)

# TL;DR

Using a single stream introduce a whole lot of painful coupling for no reason, we should use a libp2p stream handler per application.

# My proposed solution

This only affects http over stream transports.

## Get rid of  `.well-known/libp2p`

For http over stream transports `.well-known/libp2p` will not be served anywhere, instead the information is passed using the identify protocol.

## Libp2p http streams anouncements

You instead announce things with this format `/http/$PROTOCOL_ID/$HTTP_VERSION` in identify, so my example from previously, I would have theses libp2p stream handlers:
- `/http/ipfs-gateway-api/1.0`
- `/http/ipfs-gateway-api/1.1`
- `/http/ipfs-gateway-api/h2c`
- `/http/my-super-application/1.0`
- `/http/my-super-application/1.1`

For the example earlier in this document you would have thoses libp2p streams being announced:
- `/http/kad/1.0.0/1.0`
- `/http/kad/1.0.0/1.1`
- `/http/ipfs-http/1.0.0/1.0`
- `/http/ipfs-http/1.0.0/1.1`

# Other benefits:

## No weird case in case you wait for identify

This also makes code much simpler for libp2p implementations which make the choice to wait for identify to finish before opening any stream, theses implementations would either
- add a round trip before querying `.well-known/libp2p` because they wait for identify to finish
- add a weird edge case where the HTTP handling code would try to run without identify even tho the implementation might not be equipped to do theses kinds of stream opens.

## It makes it simpler to write libp2p http only implementations

Using `.well-known/libp2p` over libp2p streams is weird, you are emulating libp2p protocol discovery over libp2p, why not just use libp2p's existing infrastructure to do that instead ?
If I am implementing an new libp2p implementation from my app that purely just target http over libp2p instead of also having libp2p over http. I would need to implement the http over libp2p discovery logic even tho I don't support it.
I find that weird.

Reply:
I think the "TL;DR" section is longer than the original section 😉 .

---

> After talking with @MarcoPolo it seems that the expected minimum API for implementations would be along theses lines: ...

Just to be clear, I don't think anything in the document specifies what the API should like for implementations. I also disagree with that example, but let's save that discussion for another thread, maybe in the go-libp2p draft PR that I'll create sometime soon.

On `fasthttp`, there's nothing here that would prevent you from using `fasthttp`. In fact, you can do that today right now https://github.com/libp2p/go-libp2p-http gives you a net.Listener that you can give to `fasthttp.Serve`. Do that and you're done.

On mixing `/h2c` and `/http/1.1` with certain protocols only existing in h2c and others only in 1.1, I don't think this is a top use case we should support. Existing widely deployed HTTP systems don't need this, and we probably don't either. Unless I see a strong compelling use case, I would suggest we ignore this.

On `.well-known/libp2p` and `identify`, you can include the information in `.well-known/libp2p` inside the identify response, so there's no extra cost to this. This isn't spec'd out yet, but should be straightforward.

On `.well-known/libp2p` in general, it's very nice if the way HTTP over a libp2p stream works is identical to the way HTTP over a tls+tcp connection works. It means that a user can pass the same HTTP router (`http.ServeMux` or `http.Handler` in Go) to both an HTTP stack and libp2p stack and it would just work. We already need to define how to route a given path to a given application protocol to work on HTTP, why would we add this again via the protocol id when doing an HTTP over a libp2p stream?

Reply:
The tldr is just:

> Using a single stream introduce a whole lot of painful coupling for no reason, we should use a libp2p stream handler per underlying http service.

---

> On `fasthttp`, there's nothing here that would prevent you from using `fasthttp`. In fact, you can do that today right now https://github.com/libp2p/go-libp2p-http gives you a net.Listener that you can give to `fasthttp.Serve`. Do that and you're done.

https://github.com/libp2p/go-libp2p-http only concerns the client not the server AFAIK.

---

> On mixing /h2c and /http/1.1 with certain protocols only existing in h2c and others only in 1.1, I don't think this is a top use case we should support. Existing widely deployed HTTP systems don't need this, and we probably don't either. Unless I see a strong compelling use case, I would suggest we ignore this.

I don't understand your point, using a libp2p stream per protocol cost nothing and I IMO simplify the code because you do not have to do two services resolution (first identify then http wellknown) I can see the http wellknown being equal argument but this is only valid if you assume absolutely all libp2p http implementations would want to do both plain http and or http over libp2p streams which I don't belive to be true.

---

> On `.well-known/libp2p` and `identify`, you can include the information in `.well-known/libp2p` inside the identify response, so there's no extra cost to this. This isn't spec'd out yet, but should be straightforward.

Why not specify it now ? My spec is easy (`/http/$PROTOCOL_ID/$HTTP_VERSION`) and will not require weird backward compatible code where the information is both in identify and `.wellknown/libp2p` ?

---

> (`http.ServeMux` or `http.Handler` in Go) to both an HTTP stack and libp2p stack and it would just work.

No, what if I am using fiber which uses fasthttp ? I wont be compatible with the `http.ServeMux` nor `http.Handler`.
Go is a best case scenario, what about rust where there is no clear standard HTTP lib ?
It is not urealistical for them to have an actix handler and an hyper handler in the same app.
Using different stream handlers for different libp2p http apps cost nothing while not doing so has real costs in edge cases.

Reply:
Having everything on one single `/http/1.1` stream handler is more complex than it looks like, because with pipelinable protocols like `http/1.1` the first request can be done on application handler 1 which is actix based while the second request could be handled on a different handler backed by hyper.

Reply:
There is no pipelining here. This document says each HTTP request and response happen in a single stream. You use a libp2p stream per request.

Reply:
Actually my proposal only works with http over libp2p not libp2p over http so it's value is really limited.